docs/index-management/index-performance.md

```diff
@@ -466,15 +466,57 @@ increases. Often, such traversals can be rewritten as constrained
 traversals that can utilize a vertex-centric index to ensure acceptable
 performance at scale.
 
+### Vertex-centric indexes on implicit keys
+
+In some cases it is relevant to find an edge based on properties of the adjacent vertex.
+Let's say we want to find out whether or not Hercules has battled Cerberus.
+```groovy
+h = g.V().has('name', 'hercules').next()
+g.V(h).out('battled').has('name', 'cerberus').limit(1).hasNext()
+```
+
+A query like this can not use a vertex centric index because it filters on vertex properties rather than edge properties.
+But by restructuring the query, we can achieve exactly this.
+As both vertices are known, the vertex ids can be used to select the edge.
+In contrast to the name "Cebereus", which is a property of the adjacent vertex, the id of this vertex is already saved within the connecting edge itself.
+Although, this id is internal and therefore not explcitly visible as a property of the edge.
+This can be solved by referring to the id as an `ImplicitKey`:
+```groovy
+h = g.V().has('name', 'hercules').next()
+c = g.V().has('name', 'cerberus').next()
+g.V(h).outE('battled').has(ImplicitKey.ADJACENT_ID.name(), c.id()).limit(1).hasNext()
```

Comment:
Seeing this, I think we should implement a traversal strategy that allows JanusGraph to use this implicitly if simply `hasId()` is used. I would consider `ImplicitKey` as internal API of JanusGraph that users shouldn't directly use in their applications.

Reply:
From a user's standpoint, I think this approach is very useful and I am starting using it (In my case, JanusGraph is embedded in my application). Having a traversal strategy is very nice, but it won't hurt to have this (temporarily).

When using a gremlin query like `g.V(h).outE('battled')`, I found the response itself already contains the id of both endpoints. Since this id info is not concealed in the edge, I wouldn't really consider it as an internal API of JanusGraph.

Another concern is, `g.V(h).outE('battled').has(ImplicitKey.ADJACENT_ID.name(), c.id())` can be even faster if vertex-centric index for `ImplicitKey.ADJACENT_ID` is built, as mentioned in this PR. If users are not expected to use `ImplicitKey` directly in their applications, how could they build vertex-centric index to accelerate the traversal here?

Reply:
I'm with @li-boxuan here. Integrating ImplicitKey into hasId() is fine, but one still needs a way of creating the index in the first place. One could argue it's cleaner to provide a separate method for creating id based indexes rather than explicitly using ImplicitKey.

@FlorianHockmann I will try to figure out whether or not hasId() already profits of having an manually created index on adjacent ids. I currently doubt this is the case but I'll have a closer look on it.

Reply:
It looks like indexes are not supported by `hasId()` yet:
![edgeExistence](https://user-images.githubusercontent.com/7890887/73731285-839bfc00-4738-11ea-86e8-4ea26197c08b.png)

Reply:
@rngcntr Indeed they are not used for hasId step. Here is a related issue: #1525